src/Feature/Terminal/Feature_Terminal.re

```diff
@@ -406,3 +401,97 @@ let getLinesAndHighlights = (~colorTheme, ~terminalId) => {
      })
   |> Option.value(~default=([||], []));
 };
+
+// BUFFERRENDERER
+
+// TODO: Unify the model and renderer state. This shouldn't be needed.
+[@deriving show({with_path: false})]
+type rendererState = {
+  title: string,
+  id: int,
+  insertMode: bool,
+};
+
+let bufferRendererReducer = (state, action) => {
+  switch (action) {
+  | Service(ProcessTitleChanged({id, title, _})) when state.id == id => {
+      ...state,
+      id,
+      title,
+    }
+  | Command(NormalMode) => {...state, insertMode: false}
+  | Command(InsertMode) => {...state, insertMode: true}
+
+  | _ => state
+  };
+};
```

Comment:
BufferRenderer state and reducer has been moved here to untangle command dependencies. And also because this is where it belongs, because of dependency issues like this, but also so it can be abstracted and its implementation details hidden.